common/ofi: Add comments for the import monitor code

wckzhang opened the pull request:
Signed-off-by: William Zhang <wilzhang@amazon.com>

wckzhang commented:
Adding some comments to make the code more understandable.

bwbarrett commented on opal/mca/common/ofi/common_ofi.c:
```diff
@@ -40,6 +40,10 @@ static int opal_common_ofi_init_ref_cnt = 0;
 
+/*
+ * These no-op functions are necessary since libfabric does not allow null
+ * function pointers here or else it will segfault.
```
drop the "or else it will segfault".  The fact that it doesn't allow NULL pointers is sufficient.

bwbarrett commented on opal/mca/common/ofi/common_ofi.c:
```diff
@@ -87,19 +91,30 @@ OPAL_DECLSPEC int opal_common_ofi_init(void)
 {
```
All functions that do interesting work should have a doxygen style header.

You should call out things like what this function is doing and why.  At some point, you should explicitly say that you're injecting a monitor into Libfabric so that Libfabric can use Open MPI's memory hooks as its memory monitor, rather than trying to use its own.

wckzhang replied:
I had a comment in the header file for the function, but it isn't doxygen style, I'll update it and give it more information

bwbarrett commented on opal/mca/common/ofi/common_ofi.c:
```diff
@@ -87,19 +91,30 @@ OPAL_DECLSPEC int opal_common_ofi_init(void)
     int ret;
 
+    /*
+     * Global refcnt and boolean are used to ensure we only initialize once
```
This pattern is well established and doesn't need a comment about why its here.  However, you do need to explain how this fits in with the thread safety requirements of this function (and document those requirements in the function header).

wckzhang replied:
I think to make it easier for users to call this function, I'll add a new opal_mutex_t lock and lock around the whole function.

bwbarrett replied:
why is that the right thing to do?  That has its own risks, especially around initialization of the mutex in the initialization function you're currently writing.  Just get the definition right.

wckzhang replied:
There's already a common mutex lock that is being used, I think it's fine to re-use that

bwbarrett commented on opal/mca/common/ofi/common_ofi.c:
```diff
@@ -87,19 +91,30 @@ OPAL_DECLSPEC int opal_common_ofi_init(void)
     }
 
+    /*
+     * This cache object doesn't do much, but is necessary for the API to work.
```
this doesn't really help the reader.  Bigger questions would be "Why 1.13" and "why "mr_cache"".  Those are the things you should comment on.

bwbarrett commented on opal/mca/common/ofi/common_ofi.c:
```diff
@@ -137,6 +159,9 @@ OPAL_DECLSPEC int opal_common_ofi_init(void)
 OPAL_DECLSPEC int opal_common_ofi_fini(void)
 {
+    /*
+     * Global refcnt and boolean are used to ensure we only finalize once
```
you don't need to do this comment.

wckzhang commented:
Updated comments, added locking to the init/fini function since I don't think it was actually thread safe previously. Reused the common ofi mutex lock for this.

bwbarrett commented:
Replaced by https://github.com/open-mpi/ompi/pull/9441
